src/common/async/yield_context.h

```diff
@@ -27,27 +34,180 @@
-  operator bool() const noexcept { return y; }
+  operator bool() const noexcept {
+    return std::holds_alternative<cy>(v);
+  }
+
+  /// Bind an error code to the completion
+  optional_yield operator[](boost::system::error_code& ec) const noexcept {
+    if (*this) {
+      auto y = std::get<cy>(v).second[ec];
+      return optional_yield(get_io_context(), y);
+    } else {
+      return empty_t{}[ec];
+    }
+  }
 
   /// return a reference to the associated io_context. only valid if non-empty
-  boost::asio::io_context& get_io_context() const noexcept { return *c; }
+  boost::asio::io_context& get_io_context() const noexcept {
+    return *std::get<cy>(v).first;
+  }
 
   /// return a reference to the yield_context. only valid if non-empty
-  spawn::yield_context& get_yield_context() const noexcept { return *y; }
+  spawn::yield_context& get_yield_context() noexcept {
+    return std::get<cy>(v).second;
+  };
+
+  /// Get blocking continuation, only valid if empty.
+  ceph::async::use_blocked_t& get_blocked() noexcept {
+    return std::get<ceph::async::use_blocked_t>(v);
+  }
+
+  // Set the checker for yield contexts
+  static void set_checker(std::function<void()> f) {
+    std::unique_lock l(lock);
+    checker = std::move(f);
+  }
+
+  /// Call the checker for yield contexts
+  static void check() {
+    std::unique_lock l(lock);
+    auto f = checker;
+    l.unlock();
+    if (f) f();
+  }
 };
 
 // type tag object to construct an empty optional_yield
 static constexpr optional_yield::empty_t null_yield{};
+
+namespace boost::asio {
+template<typename ReturnType, typename... Args>
+class async_result<optional_yield, ReturnType(Args...)>
+{
+  using yield_result = async_result<spawn::yield_context, void(Args...)>;
+  using blocked_result = async_result<ceph::async::use_blocked_t,
+				       ReturnType(Args...)>;
+
+  using retvar = std::variant<yield_result, blocked_result>;
+  retvar r;
+
+public:
+  class completion_handler_type {
+    friend async_result;
+    using yield_handler =
+      typename async_result<spawn::yield_context,
+			    void(Args...)>::completion_handler_type;
+    using blocked_handler =
+      typename async_result<ceph::async::use_blocked_t,
+			    void(Args...)>::completion_handler_type;
+
+    using v_type = std::variant<yield_handler, blocked_handler>;
+    v_type v;
+
+  public:
+    completion_handler_type(optional_yield y)
+      : v(y ?
+	  v_type(yield_handler(y.get_yield_context())) :
+	  v_type(blocked_handler(y.get_blocked()))) {}
+
+    template<typename... Args2>
+    void operator ()(Args2&&... args) noexcept {
+      if (std::holds_alternative<
+	    typename async_result<ceph::async::use_blocked_t,
+	                          void(Args...)>::completion_handler_type>(v)) {
+	optional_yield::check();
+      }
+      std::visit([&](auto& h) {
+	h(std::forward<Args2>(args)...);
+      }, v);
+    }
+
+    bool yielding() const noexcept {
```

Comment:
unused? this `completion_handler_type` will be managed transparently by async initiating functions, so application code won't have the ability to call this

Reply:
Leftover from a previous version.